runtime/lua/vim/diagnostic.lua

```diff
@@ -890,22 +865,37 @@ M.handlers.signs = {
     end
 
     local ns = M.get_namespace(namespace)
-    if not ns.user_data.sign_group then
-      ns.user_data.sign_group = string.format('vim.diagnostic.%s', ns.name)
+    if not ns.user_data.sign_ns then
+      ns.user_data.sign_ns = api.nvim_create_namespace('')
     end
 
-    local sign_group = ns.user_data.sign_group
+    opts.signs.text = opts.signs.text or { 'E', 'W', 'I', 'H' }
     for _, diagnostic in ipairs(diagnostics) do
-      vim.fn.sign_place(0, sign_group, sign_highlight_map[diagnostic.severity], bufnr, {
+      local text = opts.signs.text[diagnostic.severity] and opts.signs.text[diagnostic.severity]
+        or 'U'
+      pcall(api.nvim_buf_set_extmark, bufnr, ns.user_data.sign_ns, diagnostic.lnum, 0, {
+        sign_text = text,
+        sign_hl_group = sign_highlight_map[diagnostic.severity],
         priority = get_priority(diagnostic.severity),
-        lnum = diagnostic.lnum + 1,
       })
     end
   end,
   hide = function(namespace, bufnr)
     local ns = M.get_namespace(namespace)
     if ns.user_data.sign_group and api.nvim_buf_is_valid(bufnr) then
-      vim.fn.sign_unplace(ns.user_data.sign_group, { buffer = bufnr })
+      for _, s in
+        ipairs(
+          api.nvim_buf_get_extmarks(
+            bufnr,
+            ns.user_data.sign_ns,
+            0,
+            -1,
+            { details = true, type = 'sign' }
+          )
+        )
+      do
+        pcall(api.nvim_buf_del_extmark, bufnr, ns.user_data.sign_ns, s[1])
```

Comment:
If `sign_ns` only contains sign marks, might as well use `nvim_buf_clear_namespace()` here instead of this loop.

Reply:
oh yup i forget :)